33 endpointy do pobierania dodawania opinii i komentarzy

MichLuszcz opened the pull request:
NIE Mergować dopóki nie zostaną zrobione testy jednostkowe i integracyjne przez @Sawertynn .

mGarbowski committed:
example

mGarbowski committed:
endpoint and service

mGarbowski committed:
links and authorization

MichLuszcz committed:
Added collection of user comments to User with oneToMany relation

MichLuszcz committed:
Created getUserComments, getCommentById endpoints

MichLuszcz committed:
Created commentNotFoundException

MichLuszcz committed:
Added Unique constraint to username

MichLuszcz committed:
Created getUser endpoint

MichLuszcz committed:
Added finding a comment by id or by username

MichLuszcz committed:
merge main

MichLuszcz committed:
Added new exceptions for not finding a given entity

MichLuszcz committed:
Added endpoint and service handling of adding a new comment

MichLuszcz committed:
added links to individual comments when getting comments under a review

MichLuszcz committed:
Created empty ReviewController and ReviewService

MichLuszcz committed:
Created courseNotFoundException

MichLuszcz committed:
Created ReviewService, getReviewById and getCourseReviews

MichLuszcz committed:
Created getReview, getCourseReviews and getUserReviews in ReviewController

MichLuszcz committed:
added findByCourseIdAndUsername

MichLuszcz committed:
added skeleton of deleteReview

MichLuszcz committed:
Fixed removing reviews by delete endpoint and unlinking all entities from it before deleting

MichLuszcz committed:
Changed exception thrown when user tries to delete a comment that is not his own and handled that exception

MichLuszcz committed:
Added endpoint for adding a review with authorisation check. Added authorisation check to deleting a review

MichLuszcz committed:
Added authorisation check to adding a comment

MichLuszcz committed:
Changed getting comments under a review to use username instead of userId

MichLuszcz committed:
Added links to self, reviews and comments in getUser in UserController

MichLuszcz committed:
Added CASCADE ALL to User.comments and User.reviews

MichLuszcz committed:
Changed throwing unknown exceptions that might break the app to ones the controller can handle

MichLuszcz committed:
Log which user sent the delete comment request

MichLuszcz committed:
removed cascades from User because they were casusing the deleteAll() method in User integration tests to throw a CocurrentModificationException

MichLuszcz committed:
Added todo to fix the UserPrincipal

mGarbowski commented on backend/src/main/java/pl/edu/pw/pap/comment/Comment.java:
```diff
@@ -38,6 +38,14 @@ public Comment(String text, Review review, User user) {
+    @PreRemove
+    public void removeFromUser(){
+        this.user.removeComment(this);
+        System.out.println("removed comment " + this.id + " from user");
```
Jak coś wypisujemy to przez wstrzyknięcie odpowiedniego Loggera

mGarbowski commented on backend/src/main/java/pl/edu/pw/pap/comment/commentNotFoundException.java:
```diff
@@ -0,0 +1,7 @@
+package pl.edu.pw.pap.comment;
+
+public class commentNotFoundException extends RuntimeException {
```
Nazwy klas wielką literą

mGarbowski commented on backend/src/main/java/pl/edu/pw/pap/comment/CommentService.java:
```diff
@@ -0,0 +1,83 @@
+
+        var maybeComment = commentRepository.findById(commentId);
+        if (maybeComment.isEmpty()){
+            throw new commentNotFoundException("No comment with Id" + commentId);
```
w przypadku usuwania, próba usunięcia nieistniejącego zasobu powinna zwrócić 204 No Content a nie 404 Not Found

MichLuszcz replied:
fixed

mGarbowski commented on backend/src/main/java/pl/edu/pw/pap/comment/CommentService.java:
```diff
@@ -0,0 +1,83 @@
+        if (maybeComment.isEmpty()){
+            throw new commentNotFoundException("No comment with Id" + commentId);
+        }
+        System.out.println("Trying to get user with username: " + principal.getUsername());
```
logger, nie print

mGarbowski commented on backend/src/main/java/pl/edu/pw/pap/comment/CommentService.java:
```diff
@@ -0,0 +1,83 @@
+        }
+        var comment = maybeComment.get();
+        var user = maybeUser.get();
+        if (!comment.getUser().getId().equals(user.getId())) {
```
a co jeśli jest adminem

MichLuszcz replied:
Dodałem check  żeby sprawdzał czy jeśli nie ma zgodności nazw użytkowników to czy użytkownik żądający nie jest też adminem. Jeśli nie jest to nie przechodzi dalej

mGarbowski commented on backend/src/main/java/pl/edu/pw/pap/comment/CommentService.java:
```diff
@@ -0,0 +1,83 @@
+        }
+        System.out.println("Trying to get user with username: " + principal.getUsername());
+        var maybeUser = userRepository.findByUsername(principal.getUsername());
+        if (maybeUser.isEmpty()){
```
Zamiast używać `.ifPresent()` / `.ifEmpty()` i potem `.get()` to są metody, które pozwalają bardziej elegancko to pisać jak np `.orElseThrow(...)`

mGarbowski commented on backend/src/main/java/pl/edu/pw/pap/comment/CommentService.java:
```diff
@@ -0,0 +1,83 @@
+        Long courseId = request.courseId();
+        String username = request.username();
+        Optional<User> getUser = userRepository.findByUsername(username);
+        if (getUser.isEmpty()) {
```
`.orElseThrow(...)`

mGarbowski commented on backend/src/main/java/pl/edu/pw/pap/comment/AddCommentRequest.java:
```diff
@@ -0,0 +1,4 @@
+package pl.edu.pw.pap.comment;
+
+public record AddCommentRequest(String text, Long courseId, String username) {
```
Skoro mamy `UserPrincipal` to chyba w ogóle nie potrzeba tu dawać username(?)

mGarbowski commented on backend/src/main/java/pl/edu/pw/pap/comment/CommentController.java:
```diff
@@ -0,0 +1,92 @@
+    }
+
+    @PostMapping("/api/courses/{courseId}/reviews/{username}/comments")
+    public Comment addComment(@RequestBody AddCommentRequest request, @AuthenticationPrincipal UserPrincipal principal) {
```
w tym momencie jakiś username mamy w URL, request i principal

mGarbowski commented on backend/src/main/java/pl/edu/pw/pap/course/courseNotFoundException.java:
```diff
@@ -0,0 +1,7 @@
+package pl.edu.pw.pap.course;
+
+public class courseNotFoundException extends RuntimeException {
```
Nazwy klas wielką literą

mGarbowski commented on backend/src/main/java/pl/edu/pw/pap/review/ReviewController.java:
```diff
@@ -0,0 +1,126 @@
+
+
+
+    @GetMapping("/api/courses/{courseId}/users/{username}/reviews")
```
ja bym dał `/api/courses/{courseId}/reviews/{username}`, niespójne z tym co jest niżej

MichLuszcz replied:
done

mGarbowski commented on backend/src/main/java/pl/edu/pw/pap/review/ReviewController.java:
```diff
@@ -0,0 +1,126 @@
+        // TODO: Add link from CourseController
+//        Link courseLink = linkTo(method)
+        Link[] links = {selfLink, userLink, commentsLink};
+        System.out.println("returning review with id " + review.id.courseId + " " + review.id.userId);
```
logger

mGarbowski commented on backend/src/main/java/pl/edu/pw/pap/review/ReviewController.java:
```diff
@@ -0,0 +1,126 @@
+    }
+
+
+    @PostMapping("/api/courses/{courseId}/users/{username}/reviews")
```
i tu niespójny URL

MichLuszcz replied:
zmienione na /api/courses/{courseId}/reviews/{username}

mGarbowski commented on backend/src/main/java/pl/edu/pw/pap/review/ReviewController.java:
```diff
@@ -0,0 +1,126 @@
+    }
+
+
+    @DeleteMapping("/api/courses/{courseId}/users/{username}/reviews")
```
i tu

MichLuszcz replied:
zmienione na /api/courses/{courseId}/reviews/{username}

mGarbowski commented on backend/src/main/java/pl/edu/pw/pap/review/ReviewController.java:
```diff
@@ -0,0 +1,126 @@
+
+
+
+    @ExceptionHandler({commentNotFoundException.class, userNotFoundException.class, reviewNotFoundException.class})
```
może zrobić wspólną klasę bazową EntityNotFoundException

mGarbowski commented on backend/src/main/java/pl/edu/pw/pap/user/User.java:
```diff
@@ -17,16 +18,23 @@ public class User {
     private Boolean enabled;
 
+    // TODO: figure out how to cascade reviews and comments without repository.deleteAll throwing Cocurrent Modification (if needed)
     @OneToMany(mappedBy = "user", fetch = FetchType.EAGER)
```
bez `@JsonIgnore` ?

MichLuszcz replied:
Co prawda ja tutaj akurat na tym branchu nie robiłem za bardzo, to z racji że ma przeciążony toString, i definiujemy własne endpointy, to akurat nie jest to automatycznie wrzucane do Jsonów więc nie ma potrzeby. Jedyny dostęp do tego jest przez link do innej metody kontrolera. Nie ma ryzyka pętli w aktualnej konfiguracji

mGarbowski commented on backend/src/main/java/pl/edu/pw/pap/review/ReviewService.java:
```diff
@@ -0,0 +1,85 @@
+        System.out.println("Asked for deletion of review by " + username + " of course " + courseId);
+        Optional<User> maybeUser = userRepository.findByUsername(username);
+        if (maybeUser.isPresent()){
+            System.out.println("Found user of review being deleted");
```
logger

MichLuszcz replied:
Wszystkie system.outy zostały albo zmienione na loggera albo wyrzucone całkowicie jeśli nie były specjalnie potrzebne

mGarbowski reviewed (changes requested):
Generalnie very nice,
w paru miejscach nie jestem pewien czy jest dobra logika
w wielu mi sie estetycznie nie podoba ale jest dobrze

mGarbowski committed:
some refactoring

mGarbowski committed:
Merge branch 'main' into 33-endpointy-do-pobierania-dodawania-opinii-i-komentarzy

MichLuszcz committed:
Normalised endpoint urls in ReviewController

MichLuszcz committed:
switch to streams for creating an entity model list from a list

MichLuszcz committed:
removed exception throwing from deleteComment

MichLuszcz committed:
Simplified exception throwing when found value is null

MichLuszcz committed:
Removed unecessary system.out prints

MichLuszcz committed:
Exception class names start with capital letters 🫣

MichLuszcz committed:
simplified exception throwing when entity wasn't found in ReviewController

MichLuszcz committed:
Changed name of addLinks to commentWithLinks

MichLuszcz committed:
new method for creating Reviews with links. Improved creating EntityModeLists of Reviews using streams

MichLuszcz committed:
changed CollectionModels to RepresentationModels for lists of Reviews to avoid lack of _embedded field in json

MichLuszcz committed:
cleaned up unused imports

mGarbowski committed:
use hal model builder for multiple comments

mGarbowski committed:
add criticism

MichLuszcz committed:
Changed addNewUser signature, now uses the proper users for finding the review and assigning the author of the comment

MichLuszcz committed:
Removed username from addReview . Use userPrincipal in addReview when determing the author of a new review. get course ID from url.

MichLuszcz committed:
Added link to reviews in CourseController.addLinks

MichLuszcz committed:
added links to course in appropriate ReviewController methods

mGarbowski committed:
add created timestamps to review and comment,
removed likes and dislikes from review and comment

mGarbowski committed:
make ReviewController return ReviewDTOs instead of Reviews

mGarbowski committed:
Implement CommentDTO class and use it in CommentController

A CommentDTO (Data Transfer Object) class was introduced to help in transferring specific comment data without exposing unwanted fields. All methods in the CommentController that previously returned Comment entity now return CommentDTO. Mappings and conversions have been added to the CommentService class.

MichLuszcz committed:
Merge pull request #83 from mGarbowski/create-dtos-for-comment-and-review

Introduce intermediate DTO classes

mGarbowski reviewed (approved):
ujdzie
